crates/ty_server/src/session/settings.rs

```diff
@@ -2,26 +2,40 @@ use super::options::DiagnosticMode;
 
 use ty_project::metadata::options::ProjectOptionsOverrides;
 
-/// Resolved client settings for a specific document. These settings are meant to be
-/// used directly by the server, and are *not* a 1:1 representation with how the client
-/// sends them.
-#[derive(Clone, Debug)]
-#[cfg_attr(test, derive(PartialEq, Eq))]
-pub(crate) struct ClientSettings {
-    pub(super) disable_language_services: bool,
+/// Resolved client settings that are shared across all workspaces.
+#[derive(Clone, Default, Debug, PartialEq)]
+pub(crate) struct GlobalSettings {
     pub(super) diagnostic_mode: DiagnosticMode,
+}
+
+impl GlobalSettings {
+    pub(crate) fn diagnostic_mode(&self) -> DiagnosticMode {
+        self.diagnostic_mode
+    }
+
+    pub(crate) fn combine(mut self, other: &GlobalSettings) -> GlobalSettings {
+        if self.diagnostic_mode.is_workspace() || other.diagnostic_mode.is_workspace() {
+            self.diagnostic_mode = DiagnosticMode::Workspace;
+        }
+        self
+    }
```

Comment:
I do find `combine` on the `Settings` struct a bit surprsing. The way this works in ty (and uv and ruff) is that you can combine `Options`. All options are then resolved to a single `Settings`. 


The way this could look like in your PR is that you have a `GlobalOptions` struct. Each workspace configuration contains a `GlobalOptions` field and the server can combine all of these together. `GlobalOptions` then implements an `into_settings` which resolves any default values. 

The benefit of the `Options` struct is that `diagnostic_mode` can be an `Option<DiagnosticMode>`. This allows us to distinguish between: *initialize didn't set `DiagnosticMode`* and *the user specified the default `DiagnsoticMode`*.